src/main/java/org/openmaptiles/layers/Water.java

```diff
@@ -121,11 +182,147 @@ public void process(Tables.OsmWaterPolygon element, FeatureCollector features) {
         .setAttr(Fields.INTERMITTENT, element.isIntermittent() ? 1 : 0)
         .setAttrWithMinzoom(Fields.BRUNNEL, Utils.brunnel(element.isBridge(), element.isTunnel()), 12)
         .setAttr(Fields.CLASS, clazz);
+
+      try {
+        attemptNeLakeIdMapping(element);
+      } catch (GeometryException e) {
+        e.log(stats, "omt_water",
+          "Unable to process intersections", config.logJtsExceptions());
+      }
+    }
+  }
+
+  void attemptNeLakeIdMapping(Tables.OsmWaterPolygon element) throws GeometryException {
+    // if OSM lake is too small for Z6 (e.g. area bellow ~4px) we assume there is no matching NE lake
+    var geom = element.source().worldGeometry();
+    if (geom.getArea() < OSM_ID_MATCH_AREA_LIMIT) {
+      return;
+    }
+
+    // match by name:
+    boolean match = false;
+    if (element.name() != null) {
+      for (var map : neLakeNameMaps.values()) {
+        var lakeInfo = map.get(element.name());
+        if (lakeInfo != null) {
+          match = true;
+          fillOsmIdIntoNeLake(element, geom, lakeInfo, true);
+        }
+      }
+    }
+    if (match) {
+      return;
+    }
+
+    // match by intersection:
+    List<LakeInfo> items;
+    try {
+      items = neLakeIndex.getIntersecting(geom);
+    } catch (TopologyException e) {
+      stats.dataError("omt_water_intersecting");
+      LOGGER.debug("omt_water_intersecting, OSM ID: {}",
+        element.source().id(), e);
+      final var geomFixed = GeometryFixer.fix(geom);
+      try {
+        items = neLakeIndex.getIntersecting(geomFixed);
+      } catch (TopologyException e2) {
+        throw new GeometryException("omt_water_intersecting_fix",
+          "Error getting intersecting for OSM ID " + element.source().id(), e2);
+      }
+    }
+    for (var lakeInfo : items) {
+      fillOsmIdIntoNeLake(element, geom, lakeInfo, false);
+    }
+  }
+
+  /*
+   * When we match lakes with `neLakeIndexes` then `intersetsCheckNeeded` should be `false`,
+   * otherwise `true`, to make sure we DO check the intersection but to avoid checking it twice.
+   */
+  void fillOsmIdIntoNeLake(Tables.OsmWaterPolygon element, Geometry geom, LakeInfo lakeInfo,
+    boolean intersetsCheckNeeded) throws GeometryException {
+    final Geometry neGeom = lakeInfo.geom;
+    Geometry intersection;
+    try {
+      if (intersetsCheckNeeded && !neGeom.intersects(geom)) {
+        return;
+      }
+      intersection = neGeom.intersection(geom);
+    } catch (TopologyException e) {
+      stats.dataError("omt_water_intersection");
+      LOGGER.debug("omt_water_intersection, NE ID: {}, OSM ID: {}",
```

Comment:
Maybe add a little bit more context, like "fixed invalid OSM water polygon {} when adding ID to natural earth {}"